src/azul/terra.py

```diff
@@ -387,29 +387,41 @@ class TDRClient(SAMClient):
     A client for the Broad Institute's Terra Data Repository aka "Jade".
     """
 
-    @cache
-    def lookup_source_project(self, source: TDRSourceSpec) -> str:
-        """
-        Return the name of the Google Cloud project containing the source
-        (snapshot or dataset) with the specified name.
-        """
-        return self._lookup_source(source)['dataProject']
-
-    @cache
-    def lookup_source_id(self, source: TDRSourceSpec) -> str:
-        """
-        Return the primary identifier of the source (snapshot or dataset) with
-        the specified name.
-        """
-        return self._lookup_source(source)['id']
+    @attr.s(frozen=True, kw_only=True, auto_attribs=True)
+    class TDRSource:
+        project: str
+        id: str
+        location: str
+
+    def lookup_source(self, source_spec: TDRSourceSpec) -> TDRSource:
+        source = self._lookup_source(source_spec)
+        storage = one(
+            storage
+            for dataset in (s['dataset'] for s in source['source'])
+            for storage in dataset['storage']
+            if storage['cloudResource'] == 'bigquery'
+        )
+        source = self.TDRSource(project=source['dataProject'],
+                                id=source['id'],
+                                location=storage['region'])
+        require(source.project == source_spec.project,
```

Comment:
2. Perhaps we can keep these requirements within `scripts/register_sam.py` with an added fixme to address it when/if https://github.com/DataBiosphere/azul/issues/3404 lands. Within that test that's added we can use `resolve_source` on every source in all catalogs to ensure that they are configured correctly.

Reply:
Resolved at parking lot